### ISSUE TITLE
Issues with macvlan driver & multiple networks?

### ISSUE THREAD

**Issue by h45rd** [jwilder/nginx-proxy#1050]
Hi,

Been using this great container without a hitch for quite some time. Thanks!

I recently started using the macvlan driver in addition to a user defined bridge network on certain containers and am trying to figure out if I encountered a bug or expected behavior in nginx-proxy.

Let me elaborate on how to reproduce my situation:
1. Create a container (_X_) on the user-defined bridge network nginxproxy is on
-> /etc/nginx/conf.d/default.conf is generated accordingly

2. Connect _X_ to another network using the macvlan driver
-> no changes to the default.conf

3. Restart _X_

nginx-proxy receives the event and alters **default.conf**:
```
dockergen.1 | 2018/01/18 08:17:18 Received event die for container 4560e278eb3f
dockergen.1 | 2018/01/18 08:17:18 Received event stop for container 4560e278eb3f
dockergen.1 | 2018/01/18 08:17:18 Generated '/etc/nginx/conf.d/default.conf' from 13 containers
dockergen.1 | 2018/01/18 08:17:18 Running 'nginx -s reload'
dockergen.1 | 2018/01/18 08:17:19 Received event start for container 4560e278eb3f
dockergen.1 | 2018/01/18 08:17:19 Generated '/etc/nginx/conf.d/default.conf' from 14 containers
dockergen.1 | 2018/01/18 08:17:19 Contents of /etc/nginx/conf.d/default.conf did not change. Skipping notification 'nginx -s reload'
```

-> The server line in /etc/nginx/conf.d/default.conf is altered from
`server 172.18.0.4:1234;`
to
`server 172.18.0.4 down;`
And hence the backend becomes unreachable.

Strangely I cannot always reproduce this behavior as outlined above.
Sometimes it only manifests after a restart of the docker host and/or docker daemon.

In conclusion: expected behavior or bug?

Best regards,
h45rd

**Comment by h45rd** [comment]
Forgot to mention that the affected container is always reachable from the nginx-proxy container while this occurs.

Maybe a bug in how connectivity is determined by docker-gen and/or reported by dockerd?

After some further testing, restarting the docker host with the above mentioned container setup seems to produce the described situation more often than not.

**Comment by dafal** [comment]
I have exactly the same issue with mtacvlan if I create a container from an image without the EXPOSE. EXPOSE is not needed with macvlan as the container is fully exposed.  However it seems that nginx-proxy needs it (see https://github.com/jwilder/nginx-proxy/issues/802#issuecomment-362518533).

**Comment by h45rd** [comment]
To obviate the influence of my host OS and for ease of reproduction, the following steps performed in a virgin Debian netinstall VM with only docker installed seem to reliably reproduce the above documented behavior:

```
root@debian:~# uname -a
Linux debian 4.9.0-5-amd64 #1 SMP Debian 4.9.65-3+deb9u2 (2018-01-04) x86_64 GNU/Linux
root@debian:~# docker -v
Docker version 17.12.0-ce, build c97c6d6
```

```
docker network create userdefined
docker network create -d macvlan --ip-range=192.168.0.0/24 --gateway=192.168.0.1 -o parent=ens18 macvlan
docker run -d --name airsonic --network=userdefined -e VIRTUAL_HOST=airsonic.debian.local -e VIRTUAL_PORT=4040 airsonic/airsonic
docker network connect --ip=192.168.0.100 macvlan airsonic

reboot now
```

After the host is back up, the above documented issue arises.

**Comment by h45rd** [comment]
Should I rather raise this issue with docker-gen?

**Comment by h45rd** [comment]
Any advice on how to further debug or work around this issue would be highly appreciated.
I am at my wit's end :(

**Comment by dmcgrandle** [comment]
I ran into the same issue, using macvlan driver for the "primary" Internet-facing network and bridge driver on the internal network where I put the web services containers I wanted nginx to proxy for me.  To get it to work, I followed the steps outlined for a "two container solution" with the addition of putting the jwilder/docker-gen container on the internal network.  This exposes the nginx.tmpl file, so I then edited this file.

I replaced the old line 17:
          ` server {{ .Network.IP }} down;`

with this new line:
           `server {{ .Network.IP }};`

and now it works, without adding all my containers in a down state.  Hope this helps someone else out until the experts can fix the issue.

Regards,
:D

**Comment by dmcgrandle** [comment]
Update - this was not working with containers specifying a different port (via VIRTUAL_PORT env variable), so I also modified nginx.tmpl to add a new variable called Port sent to the upstream template, line 143 from:

`{{ template "upstream" (dict "Container" $container "Address" $address "Network" $containerNetwork) }}`

to become:

`{{ template "upstream" (dict "Container" $container "Address" $address "Network" $containerNetwork "Port" $port) }}`

Now line 17 becomes:

`server {{ .Network.IP }}:{{ .Port }};`

Note: I am by no means a go template programmer, so I'm sure this will make the experts laugh.  However, the result works for me in my fairly simple environment and the macvlan networks I am trying to use with nginx-proxy.

Regards,
:D

**Comment by Archie1978** [comment]
dmcgrandle, I have the same worries as you and the pseudo solution works.
I hope that nginx-proxy will find a durable solution.

Regards

**Comment by h45rd** [comment]
Did a quick test and your workaround seems to do the trick.
Thanks, @dmcgrandle 

I would suggest a pull request, but I am not sure if the effort would be worth it.
Switched to traefik for the time being, because I gave up hope for this issue to be fixed and for some other PRs to be merged (e.g. path based routing) and did not want to keep cooking up my own patched fork.

**Comment by dmcgrandle** [comment]
Thanks guys.  Given this is just a hack to make it work, I don't think a PR would make sense.  Someone who actually knows what they are doing as a template programmer should validate this doesn't break something else important...
